tokio/examples/udp-codec.rs

```diff
@@ -0,0 +1,84 @@
+//! This example leverages `BytesCodec` to create a UDP client and server which
+//! speak a custom protocol.
+//!
+//! Here we're using the codec from tokio-io to convert a UDP socket to a stream of
+//! client messages. These messages are then processed and returned back as a
+//! new message with a new destination. Overall, we then use this to construct a
+//! "ping pong" pair where two sockets are sending messages back and forth.
+
+#![feature(async_await)]
+#![deny(warnings, rust_2018_idioms)]
+
+use env_logger;
+
+use std::net::SocketAddr;
+use std::time::Duration;
+
+use tokio;
+use tokio::io::Error;
+use tokio::net::UdpSocket;
+use tokio::util::FutureExt;
+
+#[tokio::main]
+async fn main() {
+    let _ = env_logger::init();
+
+    let addr: SocketAddr = "127.0.0.1:0".parse().unwrap();
+
+    // Bind both our sockets and then figure out what ports we got.
+    let a = UdpSocket::bind(&addr).expect("Failed to bind to address");
+    let b = UdpSocket::bind(&addr).expect("Failed to bind to address");
+    let b_addr = b.local_addr().expect("Failed to get address of socket B");
+
+    // Start off by sending a ping from a to b, afterwards we just print out
+    // what they send us and continually send pings
+    let a = pinger(a, b_addr);
+
+    // The second client we have will receive the pings from `a` and then send
+    // back pongs.
+    let b = ponger(b);
+
+    match futures::future::join(a, b).await {
+        (Err(e), _) | (_, Err(e)) => println!("an error occured; error = {:?}", e),
+        _ => println!("done!"),
+    }
+}
+
+// I moved the ping and pong routines into functions so I could use `?` for error handling.
+async fn pinger(mut socket: UdpSocket, b_addr: SocketAddr) -> Result<(), Error> {
```

Comment:
this can now be `&mut UdpSocket` since we can borrow in futures now.

Reply:
Done!